flow/DeterministicRandom.cpp

```diff
@@ -0,0 +1,122 @@
+/*
+ * DeterministicRandom.cpp
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2013-2018 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "flow/DeterministicRandom.h"
+
+uint64_t DeterministicRandom::gen64() {
+	uint64_t curr = next;
+	next = (uint64_t(random()) << 32) ^ random();
+	if (TRACE_SAMPLE()) TraceEvent(SevSample, "Random");
+	return curr;
+}
+
+DeterministicRandom::DeterministicRandom(uint32_t seed, bool useRandLog)
+  : random((unsigned long)seed), next((uint64_t(random()) << 32) ^ random()), useRandLog(useRandLog) {
+	UNSTOPPABLE_ASSERT(seed != 0); // docs for mersenne twister say x0>0
+};
+
+double DeterministicRandom::random01() {
+	double d = gen64() / double(uint64_t(-1));
+	if (randLog && useRandLog) fprintf(randLog, "R01  %f\n", d);
+	return d;
+}
+
+int DeterministicRandom::randomInt(int min, int maxPlusOne) {
+	ASSERT(min < maxPlusOne);
+	unsigned int range;
+	if (maxPlusOne < 0)
+		range = std::abs(maxPlusOne - min);
+	else {
+		range = maxPlusOne;
+		range -= min;
+	}
+	uint64_t v = (gen64() % range);
+	int i;
+	if (min < 0 && (-static_cast<unsigned int>(min + 1)) > v)
+		i = -static_cast<int>(-static_cast<unsigned int>(min + 1) - v) - 1;
+	else
+		i = v + min;
+	if (randLog && useRandLog) fprintf(randLog, "Rint %d\n", i);
+	return i;
+}
+
+int64_t DeterministicRandom::randomInt64(int64_t min, int64_t maxPlusOne) {
+	ASSERT(min < maxPlusOne);
+	uint64_t range;
+	if (maxPlusOne < 0)
+		range = std::abs(maxPlusOne - min);
+	else {
+		range = maxPlusOne;
+		range -= min;
+	}
+	uint64_t v = (gen64() % range);
+	int64_t i;
+	if (min < 0 && (-static_cast<uint64_t>(min + 1)) > v)
```

Comment:
I think we need to include equality now that we've adjusted the value by 1.

```suggestion
	if (min < 0 && (-static_cast<uint64_t>(min + 1)) >= v)
```

Reply:
Or I suppose we could not use the +1 here since this line is safe without it.